### ISSUE TITLE
[Python support] Install python files and *.pyc, *pyo object files

### ISSUE THREAD

**Issue by Salamandar** [mesonbuild/meson#2903]
I know that Python is not really supported by Meson, but with c/c++ softwares often come python bindings.
That'd be great to have some support for installing Python files with their optimized binary files with:
```meson
install_python([
        'my_files.py',
    ],
    build_pyc: true,
    build_pyo: true,
    install_dir: join_paths(mydatadir, 'python'),
)
```

Maybe that would mean building pyc/pyo at install target only, or at build time ?

**Comment by ignatenkobrain** [comment]
note that distributions explicitly do not want this as this slows down building packages because distro regenerates pyc and pyo itself.

**Comment by Salamandar** [comment]
Oh ? Well I'm currently building Gimp and makepkg/pacman do not regenerate those files.

**Comment by Salamandar** [comment]
Autotools [provides this Python compilation](https://www.gnu.org/software/automake/manual/html_node/Python.html). That's how Gimp procedes.

**Comment by TingPing** [comment]
It should behave like most post-install steps, if `DESTDIR` is set (aka a distro package) don't generate them, otherwise do. I think pacman not having a rule for that is the odd one out but they probably do that because the other tooling already did it.

In the short term you make a post-install script: https://github.com/pithos/pithos/blob/master/meson_post_install.py#L20-L21

In the long term this could be handled automatically in a method similar to my commit in #837

**Comment by Salamandar** [comment]
> if DESTDIR is set (aka a distro package) don't generate them, otherwise do

Why is that ? makepkg sets DESTDIR, but I want them to be generated.

> In the short term you make a post-install script

Looks okay for me.

**Comment by TingPing** [comment]
> Why is that ? makepkg sets DESTDIR, but I want them to be generated.

As mentioned I think Arch is the only distro that doesn't do this themselves. OpenSUSE, Debian/Ubuntu, Fedora, etc all build bytecode themselves.

**Comment by Salamandar** [comment]
> As mentioned I think Arch is the only distro that doesn't do this themselves. OpenSUSE, Debian/Ubuntu, Fedora, etc all build bytecode themselves.
Oh, okay.

**Comment by eli-schwartz** [comment]
> It should behave like most post-install steps, if `DESTDIR` is set (aka a distro package) don't generate them, otherwise do. I think pacman not having a rule for that is the odd one out but they probably do that because the other tooling already did it.

Wrong. We don't have a rule for it, because as a KISS distro we believe that this is upstream's job, and not something that should be bluntly hacked around in our packaging scripts. :stuck_out_tongue:

makepkg uses a simple bash script with some pkgname, pkgver, etc. metadata which defines `build()` and `package()` functions that the package maintainer uses to call the upstream build commands. Unlike many other distros, we do not provide "magic" wrappers for autoreconf, ./configure, make, setup.py, cmake, or anything else. We have no plans to provide magic wrappers for byte-compiling python files either, and if we absolutely must then each and every Arch package will end up running `meson ..; ninja; DESTDIR="$pkgdir" ninja install` and then afterwards run `python -m compileall "$pkgdir"/path/to/python3/files`.

It would be nice if meson followed in the footsteps of setuptools in this respect. At the very least, for the sake of people who do *not* set `DESTDIR` and do not have Debian/Fedora/SUSE doing this for them... and once the functionality exists at all, then it should be trivial to expose this via a meson option in addition/instead.

Things like `update-desktop-database`, `update-mime-database`, etc are *true* post-install steps, as they are very dependent on the live OS. (In fact, you cannot run them beforehand because they write to the same file for every piece of software.)
In comparison, byte-compiling pyc files has nothing intrinsically to do with post-install, but many distros implement it that way regardless for example to reduce package size.

**Comment by TingPing** [comment]
@eli-schwartz It isn't remotely a hack, plenty of tasks require post steps in packages. The benefit for the distro doing it is the same as setting optimization levels when compiling C, it gives them control. (I realize Arch doesn't use Meson's --buildtype=plain either though)

**Comment by eli-schwartz** [comment]
@heftig, @anthraxx, I'm not sure why we actually have this helper I didn't realize we had, since we've never shipped an arch-configure, arch-cmake, or arch-autogen.sh either. But @TingPing raises a good point that this definitely should not be setting the buildtype since makepkg.conf already has CFLAGS/CXXFLAGS for this purpose, and this would lead to an inconsistent experience when locally rebuilding packages for debug info. -_-

**Comment by dzhang-b** [comment]
hey, `install_python` target will be helpful, especially take cares of variable like `pyexecdir`, `pythondir`, `pkgpyexecdir`, `pkgpythondir` in autotools.

**Comment by mgorny** [comment]
> note that distributions explicitly do not want this as this slows down building packages because distro regenerates pyc and pyo itself.

There are distros that do want this, and having the byte-compilation levels configurable would be the best option for Gentoo (much like `python -m installer`'s `--compile-bytecode` option).

**Comment by nanonyme** [comment]
I would note that even with setuptools it's not setuptools that creates bytecode these days. You instead have pip invoke setuptools backend which generates a wheel *without any bytecode*, then installing said wheel generates bytecode in a way that completely ignores package was a setuptools package.

**Comment by eli-schwartz** [comment]
This was implemented in #11530 including a meson option to disable it.
